pkg/agent/openflow/pipeline.go

```diff
@@ -624,6 +632,112 @@ func (c *client) localProbeFlow(localGatewayIP net.IP, category cookie.Category)
+		// characters: 1) the ct_state is "+new+trk", 2) reg1[17] is set as 1; 3) Node IP is used as the target
+		// source IP in NAT action, 4) ct_mark is set with 0x40 in the conn_track context.
+		c.pipeline[conntrackCommitTable].BuildFlow(priorityNormal).
+			MatchProtocol(binding.ProtocolIP).
+			MatchCTStateNew(true).MatchCTStateTrk(true).
+			MatchRegRange(int(marksReg), snatRequiredMark, snatMarkRange).
+			Action().CT(true, l2ForwardingOutTable, ctZone).
+			SNAT(snatIPRange, nil).
+			LoadToMark(snatCTMark).CTDone().
+			Cookie(c.cookieAllocator.Request(category).Raw()).
+			Done(),
+	}
+	return flows
+}
+
+func (c *client) l3ToExternalFlows(nodeIP net.IP, localSubnet net.IPNet, outputPort int, category cookie.Category) []binding.Flow {
+	flows := []binding.Flow{
+		// Resubmit the packet to L2ForwardingCalc table if it is communicating to a Service.
+		c.pipeline[l3ForwardingTable].BuildFlow(priorityNormal).
+			MatchProtocol(binding.ProtocolIP).
+			MatchRegRange(int(marksReg), markTrafficFromLocal, binding.Range{0, 15}).
+			MatchCTMark(gatewayCTMark).
+			Action().ResubmitToTable(l2ForwardingCalcTable).
+			Cookie(c.cookieAllocator.Request(category).Raw()).
+			Done(),
+		// Resubmit the packet to L2ForwardingCalc table if it is sent to the Node IP(not to the host gateway). Since
+		// the packet is using the host gateway's MAC as dst MAC, it will be sent out from "gw0". This flow entry is to
+		// avoid SNAT on such packet, otherwise the source and destination IP are the same.
+		c.pipeline[l3ForwardingTable].BuildFlow(priorityLow).
+			MatchProtocol(binding.ProtocolIP).
+			MatchRegRange(int(marksReg), markTrafficFromLocal, binding.Range{0, 15}).
+			MatchDstIP(nodeIP).
+			Action().ResubmitToTable(l2ForwardingCalcTable).
+			Cookie(c.cookieAllocator.Request(category).Raw()).
+			Done(),
+		// Resubmit the packet to L2ForwardingCalc table if it is a packet sent to a local Pod. This flow entry has a
+		// low priority to avoid overlapping with those packets received from tunnel port.
+		c.pipeline[l3ForwardingTable].BuildFlow(priorityLow).
+			MatchProtocol(binding.ProtocolIP).
+			MatchRegRange(int(marksReg), markTrafficFromLocal, binding.Range{0, 15}).
+			MatchDstIPNet(localSubnet).
+			Action().ResubmitToTable(l2ForwardingCalcTable).
+			Done(),
+		// Add SNAT mark on the packet that is not filtered by other flow entries in L3Forwarding table. This is the
+		// table miss if SNAT feature is enabled.
+		c.pipeline[l3ForwardingTable].BuildFlow(prioritySNAT).
+			MatchProtocol(binding.ProtocolIP).
+			MatchRegRange(int(marksReg), markTrafficFromLocal, binding.Range{0, 15}).
+			Action().LoadRegRange(int(marksReg), snatRequiredMark, snatMarkRange).
+			Action().ResubmitToTable(ingressRuleTable).
+			Cookie(c.cookieAllocator.Request(category).Raw()).
+			Done(),
+		// Output the SNATed packet to the specified port.
+		c.pipeline[l2ForwardingOutTable].BuildFlow(priorityNormal).
```

Comment:
I used wanted to output the SNATed packets to the uplink port directly, so I added this flow. But later I found that if I did so, there might existed some issues, like it would be difficult to consume the host networking stack to decide the dst MAC. So I still output the packet to gw0 finally, but I didn't remove this one. You are correct, actually we could remove this one and leverage the result in  l2ForwardingCalcTable.